src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -0,0 +1,229 @@
+            // happen looking at just model metadata.
+            var key = new Key(providerContext.Metadata, token);
+
+            // If we're currently recursively building a binder for this type, just return
+            // a DelegatingBinder. We'll fix it up later to point to the 'real' binder
+            // when the stack unwinds.
+            var stack = providerContext.Stack;
+            for (var i = 0; i < stack.Count; i++)
+            {
+                var entry = stack[i];
+                if (key.Equals(entry.Key))
+                {
+                    if (entry.Value == null)
+                    {
+                        // Recursion detected, create a DelegatingBinder.
+                        var binder = new DelegatingBinder();
+                        stack[i] = new KeyValuePair<Key, DelegatingBinder>(entry.Key, binder);
+                        return binder;
+                    }
+                    else
+                    {
+                        return entry.Value;
+                    }
+                }
+            }
+
+
+            // OK this isn't a recursive case (yet) so "push" an entry on the stack and then ask the providers
+            // to create the binder.
+            stack.Add(new KeyValuePair<Key, DelegatingBinder>(key, null));
+
+            IModelBinder result = null;
+
+            for (var i = 0; i < _providers.Length; i++)
+            {
+                var factory = _providers[i];
+                result = factory.Create(providerContext);
+                if (result != null)
+                {
+                    break;
+                }
+            }
+
+            if (result == null && stack.Count > 1)
+            {
+                // Use a no-op binder if we're below the top level. At the top level, we throw.
+                result = NoOpBinder.Instance;
+            }
+            
+            // "pop"
+            var delegatingBinder = stack[stack.Count - 1].Value;
+            stack.RemoveAt(stack.Count - 1);
+
+            // If the DelegatingBinder was created, then it means we recursed. Hook it up to the 'real' binder.
+            if (delegatingBinder != null)
+            {
+                delegatingBinder.Inner = result;
+            }
+
+            return result;
+        }
+
+        private class DefaultModelBinderProviderContext : ModelBinderProviderContext
+        {
+            private readonly ModelBinderFactory _factory;
+
+            public DefaultModelBinderProviderContext(
+                ModelBinderFactory factory,
+                ModelBinderFactoryContext factoryContext)
+            {
+                _factory = factory;
+                Metadata = factoryContext.Metadata;
+                BindingInfo = factoryContext.BindingInfo;
+
+                MetadataProvider = _factory._metadataProvider;
+                Stack = new List<KeyValuePair<Key, DelegatingBinder>>();
+            }
+
+            public DefaultModelBinderProviderContext(
```

Comment:
Perhaps `private` to make it more clear where this constructor overload is used.